src/pg_copy.c

```diff
@@ -0,0 +1,552 @@
+		rightOp = get_rightop((Expr *) equalityExpr);				 
+		Assert(IsA(rightOp, Const));
+		rightConst = (Const *) rightOp;
+		rightConst->constvalue = 0;
+		rightConst->constisnull = false;
+		rightConst->constbyval = get_typbyval(columnOid);
+		whereClauseList = list_make1(equalityExpr);
+
+		/* allocate column values and nulls arrays */
+		rel = heap_open(tableId,  AccessShareLock);
+		tupleDescriptor = RelationGetDescr(rel);
+		columnCount = tupleDescriptor->natts;
+		columnValues = palloc0(columnCount * sizeof(Datum));
+		columnNulls = palloc0(columnCount * sizeof(bool));
+
+		/*
+		 * We create a new memory context called tuple context, and read and write
+		 * each row's values within this memory context. After each read and write,
+		 * we reset the memory context. That way, we immediately release memory
+		 * allocated for each row, and don't bloat memory usage with large input
+		 * files.
+		 */
+		tupleContext = AllocSetContextCreate(CurrentMemoryContext,
+											 "COPY Row Memory Context",
+											 ALLOCSET_DEFAULT_MINSIZE,
+											 ALLOCSET_DEFAULT_INITSIZE,
+											 ALLOCSET_DEFAULT_MAXSIZE);
+				
+		/* 
+		 * Construct hash table used for shardId->Connection mapping.
+		 * We can not use connection cache from connection.c used by GteConnection because
+		 * we need to establish multiple connections with each nodes: one connection per shard
+		 */
+		shardToConn = CreateShardToConnectionHash();
+
+		/* init state to read from COPY data source */
+		copyState = BeginCopyFrom(rel, copyStatement->filename,
+								  copyStatement->is_program,
+								  copyStatement->attlist,
+								  copyStatement->options);
+				
+		/* TODO: handle binary mode */
+		if (copyState->binary) { 
+			elog(ERROR, "Copy in binary mode is not currently supported");
+		}
+
+		PG_TRY();
+		{
+			/* Lock all shards in shared mode */
+			shardIntervalList = SortList(shardIntervalList, CompareTasksByShardId);
+			foreach(shardIntervalCell, shardIntervalList)
+			{
+				ShardInterval *shardInterval = (ShardInterval *) lfirst(shardIntervalCell);
+				ShardId shardId = shardInterval->id;
+				LockShardData(shardId, ShareLock);
+				LockShardDistributionMetadata(shardId, ShareLock);
+			}
+
+			while (nextRowFound)
+			{
+				MemoryContext oldContext = MemoryContextSwitchTo(tupleContext);
+				nextRowFound = NextCopyFrom(copyState, NULL, columnValues, columnNulls, NULL);
+				MemoryContextSwitchTo(oldContext);
+					
+				/* write the row to the shard */
+				if (nextRowFound)
+				{	
+					if (columnNulls[partitionColumn->varattno]) 
+					{ 
+						ereport(ERROR, (errcode(ERRCODE_NULL_VALUE_NOT_ALLOWED),
+										errmsg("cannot copy row with NULL value "
+											   "in partition column")));
+					} 
+					rightConst->constvalue = columnValues[partitionColumn->varattno];
+					prunedList = PruneShardList(tableId, whereClauseList, shardIntervalList);
+					
+					foreach(shardIntervalCell, prunedList)
```

Comment:
Instead of a foreach loop, this should be something like:
shardInterval = (ShardInterval *) linitial(prunedList);

We should copy to at most one matching shard.